flink-connector-examples/src/main/java/io/pravega/example/flink/streamcuts/process/StreamBookmarker.java

```diff
@@ -150,46 +145,50 @@ public void open(Configuration parameters) {
      */
     @Override
     public void processElement(Tuple2<Integer, Double> value, Context ctx, Collector<SensorStreamSlice> out) throws IOException {
-        if (lastValue.value() != null) return;
-        // Infer the slice of the stream for which events are < 0.
-        if (value.f1 < 0 && startStreamCut.value() == null) {
-            startStreamCut.update(getReaderGroup().getStreamCuts().get(Stream.of(Constants.DEFAULT_SCOPE, Constants.PRODUCER_STREAM)));
-            LOG.warn("Start bookmarking a stream slice at: {} for sensor {}.", startStreamCut.value(), value.f0);
-        } else if (value.f1 >= 0 && startStreamCut.value() != null && lastStreamCutBeforeSlice.value() == null) {
-            lastStreamCutBeforeSlice.update(getReaderGroup().getStreamCuts().get(Stream.of(Constants.DEFAULT_SCOPE, Constants.PRODUCER_STREAM)));
-            lastValue.update(value);
-            long sliceCreationTime = ctx.timerService().currentProcessingTime() + StreamBookmarker.CHECKPOINT_INTERVAL;
-            LOG.warn("Found events > 0 for sensor {}, waiting for the next stream cut update to create the slice at {}.", value.f0, sliceCreationTime);
-            ctx.timerService().registerProcessingTimeTimer(sliceCreationTime);
+        // Initialize task state for pending slices.
+        if (pendingBookmarks.value() == null) {
+            pendingBookmarks.update(new HashMap<>());
+        }
+
+        if (value.f1 < 0 && !pendingBookmarks.value().containsKey(value.f0)) {
+            // Instantiate a SensorStreamSlice object for this sensor.
+            SensorStreamSlice sensorStreamSlice = new SensorStreamSlice(value.f0);
+
+            // Set the current ReaderGroup StreamCut as the beginning of the slice of events of interest.
+            StreamCut startStreamCut = getReaderGroup().getStreamCuts().get(Stream.of(Constants.DEFAULT_SCOPE, Constants.PRODUCER_STREAM));
+            sensorStreamSlice.setStart(startStreamCut);
+            Map<Integer, SensorStreamSlice> currentPendingBookmarks = pendingBookmarks.value();
+            currentPendingBookmarks.put(value.f0, sensorStreamSlice);
+            pendingBookmarks.update(currentPendingBookmarks);
+            LOG.warn("Start bookmarking a stream slice at: {} for sensor {}.", startStreamCut, value.f0);
+        } else if (value.f1 >= 0 && pendingBookmarks.value().containsKey(value.f0)) {
+            SensorStreamSlice sensorStreamSlice = pendingBookmarks.value().get(value.f0);
+            StreamCut currentStreamCut = getReaderGroup().getStreamCuts().get(Stream.of(Constants.DEFAULT_SCOPE, Constants.PRODUCER_STREAM));
+
+            // If this is the first event > 0, then we need to keep the current StreamCut to check for the next one.
```

Comment:
I don't understand the logic here.   Isn't the `currentStreamCut ` a possible candidate for the end of the slice?   Seems like this logic skips over at least one checkpoint.